src/compiler/factory/node_factory.rs

```diff
@@ -0,0 +1,50 @@
+use crate::{
+    BaseNode, BaseNodeFactory, EmptyStatement, NodeArray, NodeArrayOrVec, NodeFactory, SourceFile,
+    SyntaxKind,
+};
+
+impl NodeFactory {
+    pub fn create_node_array<TElements: Into<NodeArrayOrVec>>(
```

Comment:
Rather than having the intermediate `Into<NodeArrayOrVec>`, could you have `TElements: Into<NodeArray>`? `NodeArray` could implement `From<NodeArray>` (if that's not automatic) and `From<Vec<Node>>`.